tests/unit/s2n_validate_x509_time_test.c

```diff
@@ -0,0 +1,247 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+bool s2n_libcrypto_supports_flag_no_check_time();
+uint64_t s2n_libcrypto_awslc_api_version(void);
+
+static uint8_t s2n_verify_host_accept_everything(const char *host_name, size_t host_name_len, void *data)
+{
+    return 1;
+}
+
+int main(int argc, char *argv[])
+{
+    BEGIN_TEST();
+
+    /* s2n_config_validate_x509_time tests */
+    {
+        /* Safety */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_config_validate_x509_time(NULL, true), S2N_ERR_NULL);
+        EXPECT_FAILURE_WITH_ERRNO(s2n_config_validate_x509_time(NULL, false), S2N_ERR_NULL);
+
+        /* Ensure s2n_config_validate_x509_time sets the proper state */
+        {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_EQUAL(config->validate_x509_time, true);
+
+            EXPECT_SUCCESS(s2n_config_validate_x509_time(config, false));
+            EXPECT_EQUAL(config->validate_x509_time, false);
+
+            EXPECT_SUCCESS(s2n_config_validate_x509_time(config, true));
+            EXPECT_EQUAL(config->validate_x509_time, true);
+        }
+    }
+
+    /* Test the NO_CHECK_TIME flag feature probe. The flag was added to AWS-LC in API version 19. */
+    if (s2n_libcrypto_is_awslc() && s2n_libcrypto_awslc_api_version() > 19) {
+        EXPECT_TRUE(s2n_libcrypto_supports_flag_no_check_time());
+    }
+
+    /* Test disabling x509 time validation.
+     *
+     * By default, validation should fail for certificates with invalid timestamps. However, if
+     * x509 time validation is disabled, validation should succeed.
+     *
+     * When time validation is disabled, s2n_config_set_wall_clock() will not set a custom time on
+     * the libcrypto, so this function cannot be used to set a fake time for testing. Instead, the
+     * test certificates themselves contain invalid timestamps.
+     */
+    {
+        /* clang-format off */
+        struct {
+            const char *cert_pem_path;
+            const char *key_pem_path;
+            bool validate_x509_time;
+            s2n_error expected_error;
+        } test_cases[] = {
```

Comment:
Maybe add a test for a cert that should fail validation for a reason unrelated to time, to make sure that's unaffected.